File: Data Structures & Algorithms/copy-linked-list-with-random-pointer/submission-2.java
/*
// Definition for a Node.
class Node {
    int val;
    Node next;
    Node random;

    public Node(int val) {
        this.val = val;
        this.next = null;
        this.random = null;
    }
}
*/

class Solution {
    public Node copyRandomList(Node head) {
        Node prev = new Node(0);
        Node curr = prev;
        Map<Node, Node> map = new HashMap<>(); // maps orignal -> copy
        while (head!=null) {
            Node copy = map.getOrDefault(head, new Node(head.val));
            map.put(head, copy);
            curr.next = copy;
            curr = curr.next; // now at. copy
            if (head.random != null) {
                copy = map.getOrDefault(head.random, new Node(head.random.val));
                map.put(head.random, copy);
                curr.random = copy;
            }
            head = head.next;
        }
        return prev.next;
    }
}

File: Data Structures & Algorithms/reverse-linked-list-ii/submission-2.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverseBetween(self, head: Optional[ListNode], left: int, right: int) -> Optional[ListNode]:
        index = 0
        dummy = ListNode(0, head)
        maps = {index: dummy}
        node = head
        while node:
            index += 1
            maps[index] = node
            node = node.next
        maps[index+1] = node
        if index == 1:
            return head
        prev = None
        node = maps[left]
        while node != maps[right+1]:
            temp = node.next
            node.next = prev
            prev = node
            node = temp
        maps[left-1].next = maps[right]
        maps[left].next = maps[right+1]
        return dummy.next

        